`TypeError`/`SystemError` for `dataclasses` that uses `default_factory` and is subclassed

```python
from importlib.metadata import version
from dataclasses import dataclass, field


@dataclass
class _Test:
    x: dict = field(default_factory=dict)


class Test(_Test):
    pass


print("Mypy version:", version("mypy"))
print(Test())
```

With `mypyc test.py && python -c "import test"` leads to:
```shell
❯ mypyc test.py && python -c "import test"
test.py:7: warning: Unsupported default attribute value
running build_ext
building 'test' extension
clang -fno-strict-overflow -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -O3 -Wall -I/private/tmp/mypyc/.venv/lib/python3.14/site-packages/mypyc/lib-rt -I/private/tmp/mypyc/.venv/include -I/opt/homebrew/opt/python@3.14/Frameworks/Python.framework/Versions/3.14/include/python3.14 -c build/__native.c -o build/temp.macosx-26.0-arm64-cpython-314/build/__native.o -O3 -g1 -Werror -Wno-unused-function -Wno-unused-label -Wno-unreachable-code -Wno-unused-variable -Wno-unused-command-line-argument -Wno-unknown-warning-option -Wno-unused-but-set-variable -Wno-ignored-optimization-argument -Wno-array-bounds -Wno-stringop-overread -Wno-stringop-overflow -Wno-cpp -Wno-array-bounds -Wno-stringop-overread -Wno-stringop-overflow
clang -bundle -undefined dynamic_lookup build/temp.macosx-26.0-arm64-cpython-314/build/__native.o -o build/lib.macosx-26.0-arm64-cpython-314/test.cpython-314-darwin.so
copying build/lib.macosx-26.0-arm64-cpython-314/test.cpython-314-darwin.so -> 
Mypy version: 2.3.0
TypeError: dict object expected; got dataclasses.Field

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
    import test
  File "test.py", line 15, in <module>
    print(Test())
  File "<string>", line 3, in __init__
SystemError: <class 'dict'> returned a result with an exception set
```

Ran into this in the wild while trying to compile `isort` with `mypyc` (https://github.com/PyCQA/isort/pull/2586).

Code runs fine without `mypyc` compilation.

--- a/mypyc/irbuild/classdef.py
+++ b/mypyc/irbuild/classdef.py
@@ -760,7 +760,6 @@ def find_attr_initializers(
     if cls.builtin_base:
         return set(), []
 
-    cls_type = dataclass_type(cdef)
     attrs_with_defaults: set[str] = set()
     default_assignments: list[tuple[AssignmentStmt, str]] = []
 
@@ -774,10 +773,11 @@ def find_attr_initializers(
         info_ir = builder.mapper.type_to_ir.get(info)
         if info_ir is None:
             continue
+        info_cls_type = dataclass_type(info.defn)
         for stmt in info.defn.defs.body:
             if not isinstance(stmt, AssignmentStmt):
                 continue
-            name = default_attr_name(stmt, info_ir, cls_type)
+            name = default_attr_name(stmt, info_ir, info_cls_type)
             if name is None:
                 continue
             attrs_with_defaults.add(name)

--- a/mypyc/test-data/run-python37.test
+++ b/mypyc/test-data/run-python37.test
@@ -74,8 +74,15 @@ class Person5:
     friends: Set[str] = field(default_factory=set)
     parents: FrozenSet[str] = frozenset()
 
+@dataclass
+class HasFactoryDefault:
+    x: dict[str, int] = field(default_factory=dict)
+
+class InheritsDataclassInit(HasFactoryDefault):
+    pass
+
 [file other.py]
-from native import Person1, Person1b, Person2, Person3, Person4, Person5, testBool
+from native import Person1, Person1b, Person2, Person3, Person4, Person5, testBool, InheritsDataclassInit
 i1 = Person1(age = 5, name = 'robot')
 assert i1.age == 5
 assert i1.name == 'robot'
@@ -125,6 +132,9 @@ assert Person1.__annotations__ == {'age': int, 'name': str}
 assert Person2.__annotations__ == {'age': int, 'name': str}
 assert Person5.__annotations__ == {'weight': float, 'friends': set,
                                        'parents': frozenset}
+v = InheritsDataclassInit()
+assert isinstance(v.x, dict)
+assert v.x == {}
 
 [file driver.py]
 import sys

